src/gil.rs

```diff
@@ -137,7 +136,56 @@ impl GILGuard {
     /// If PyO3 does not yet have a `GILPool` for tracking owned PyObject references, then this
     /// new `GILGuard` will also contain a `GILPool`.
     pub(crate) fn acquire() -> GILGuard {
-        prepare_freethreaded_python();
+        // Maybe auto-initialize the GIL:
+        //  - If auto-initialize feature set and supported, try to initalize the interpreter.
+        //  - If the auto-initialize feature is set but unsupported, emit hard errors only when
+        //    the extension-module feature is not activated - extension modules don't care about
+        //    auto-initialize so this avoids breaking existing builds.
+        //  - Otherwise, just check the GIL is initialized.
+        cfg_if::cfg_if! {
+            if #[cfg(all(feature = "auto-initialize", Py_SHARED, not(PyPy)))] {
+                prepare_freethreaded_python();
+            } else if #[cfg(all(feature = "auto-initialize", not(feature = "extension-module"), not(Py_SHARED), not(rustdoc)))] {
```

Comment:
I think we can make this simpler by `else if all(auto-initialize, PyPy) { ... } else if all(auto_initialize, ...)`

Reply:
`not(extension-module)` is unfortunately necessary to avoid a breaking change - otherwise all extension modules would suddenly start failing. We could make it as you suggest in `0.14`.

Reply:
Ah, thanks. That makes sense.
BTW, do you think that we should keep auto-initialize in default-features?

Reply:
I think in the long-run it's better to remove `auto-initialize` from default-features, but that's also a breaking change, so let's remove it in `0.14`.

I'll create an issue to track these cleanups and add it to the `0.14` milestone.

Reply:
👍🏼